src/resolver/nodeModuleLookup.ts

```diff
@@ -36,29 +36,32 @@ export function isNodeModule(path: string): undefined | IModuleParsed {
   return result;
 }
 
-export function parseAllModulePaths(fileAbsPath: string) {
+export function parseAllModulePaths(fileAbsPath: string): string[] {
   const baseDir = path.dirname(fileAbsPath);
+  const segments = baseDir.split(path.sep).filter(s => s);
   const paths = [];
+  for (let i = segments.length; i >= 0; i--) {
+    if (i > 0 && segments[i - 1] === "node_modules")
+      continue;
+    const candidate = path.join("/", segments.slice(0, i).join(path.sep), "node_modules");
+    paths.unshift(candidate);
+  }
 
-  const snippets = baseDir.split(/node_modules/);
-  let current = '';
-  if (snippets.length > 1) {
-    const total = snippets.length - 1;
-    for (let i = 0; i < total; i++) {
-      current = path.join(current, snippets[i], 'node_modules');
-      paths.push(current);
-    }
-
-    const last = snippets[total];
+  return paths;
+}
 
-    const matchedLast = last.match(/[\/|\\]([a-z-@0-9_-]+)/gi);
-    if (matchedLast && matchedLast[0]) {
-      paths.push(path.join(paths[paths.length - 1], matchedLast[0], 'node_modules'));
-    }
-    return paths;
+const pathExists = new Map<string, boolean>();
+function memoizedExists(absPath: string): boolean {
+  let exists = pathExists.get(absPath);
+  if (exists === undefined) {
+    exists = fileExists(absPath);
+    pathExists.set(absPath, exists);
   }
+  return exists;
+}
 
-  return [PROJECT_NODE_MODULES];
+export function parseExistingModulePaths(fileAbsPath: string): string[] {
+  return parseAllModulePaths(fileAbsPath).filter(memoizedExists);
```

Comment:
That looks legit, except `filter` is very slow. If you noticed we're trying to avoid using those function in favour of for loops;-)
https://jsperf.com/for-vs-filter

That's a nitpick tho. If you can fix it that would be great ;-)